src/assisted_test_infra/test_infra/utils/utils.py

```diff
@@ -617,3 +649,67 @@ def get_iso_download_path(entity_name: str):
 def get_major_minor_version(openshift_full_version: str) -> str:
     semantic_version = semver.VersionInfo.parse(openshift_full_version, optional_minor_and_patch=True)
     return f"{semantic_version.major}.{semantic_version.minor}"
+
+
+def waiter_decorator(func):
+    """decorate cluster waiters
+    When waiter called from cluster timeout exception may raise without information
+    Adding information about the timeout from job.batch/assisted-installer-controller,
+    Taking the last error log levels and append to the timeout message object.
+
+    We will be able to download kubeconfig in the next states:
+    installing, finalizing, installed, error, adding-hosts, cancelled, installing-pending-user-action
+
+    Sometimes logs from job.batch/assisted-installer-controller may return empty or failed when:
+    - In case pod/assisted-installer-controller-zzz in pending status (~10 minutes after installing phase)
+    - In case api vip address not accessible.
+
+    Current code will retry_success 3 times log assisted-installer-controller.
+    The code will ignore failures when trying to get assisted-installer-controller logs  and return
+    the timeout exception with pre-configured message
+    """
+
+    def update_kube_config(cluster_object):
+        # download kubeconfig and update /etc/hosts with api-vip host ip
+        try:
+            cluster_object.download_kubeconfig()
+            config_etc_hosts(
+                cluster_name=cluster_object.name,
+                base_dns_domain=cluster_object.get_details().base_dns_domain,
+                api_vip=cluster_object.get_details().api_vips[0].ip,
+            )
+            return True
+        except Exception:
+            log.info(f"Unable to download kubeconfig from cluster {cluster_object.name} - ignored")
+            return False
+
+    def filter_log_by(output, filter_by, repeat_count=2):
+        if output:
+            lines = output.splitlines()
+            reverse_list = [line + "\n" for line in lines[::-1] if filter_by in line]
+            str_represent = map(str, reverse_list[:repeat_count])
+            str_represent = "\n".join(str_represent)
+            log.info(f"assister-installer-controller filter: {str_represent}")
+            return str_represent
+        log.info("assister-installer-controller filter returned empty")
+        return ""
```

Comment:
can you explain what this function does with an example?

Reply:
Once we get timeout exception from waiter, we pass output to  filter_log_by func.
 the output from the command (the logs) and extract filter_by string we are looking for . the result appended to the timeout exceptions.

Example:
In Ci runs we got back on timeout installation :
Error Message
`waiting.exceptions.TimeoutExpired: Timeout of 3600 seconds expired waiting for Nodes to be in of the statuses ['installed']`

This message has nothing about the issue , it might be operators timeout or any other . the idea is to append to this message the "error" from job.batch/assisted-installer-controller 

So the filtered called , 
getting log as a list , reverse it because we want the 2 last filter_by we found.  in case filter is "error=" 
split the lines by /n , 
get all lines with "error="  , reverse the list because we need only the last messages.
convert the list to str and return .

Example:
the output is :

```
> ............
> time="2024-01-14T03:28:31Z" level=error msg="Failed to get list of CSRs." func=github.com/openshift/assisted-installer/src/k8s_client.k8sClient.ListCsrs file="/remote-source/app/src/k8s_client/k8s_client.go:367" error="the server was unable to return a response in the time allotted, but may still be processing the request (get certificatesigningrequests.certificates.k8s.io)"
> time="2024-01-14T03:28:31Z" level=error msg="Failed to get list of nodes from k8s client" func="github.com/openshift/assisted-installer/src/assisted_installer_controller.(*controller).waitAndUpdateNodesStatus" file="/remote-source/app/src/assisted_installer_controller/assisted_installer_controller.go:248" error="the server was unable to return a response in the time allotted, but may still be processing the request (get nodes)" request_id=191ba4cd-051f-41c5-9969-d5f90367b11e
> ...................
```

All sent to filter_by as outpu , looking for  "error="  lines and stored the last 2  (reverst listed)
convert the list element to str.
append the messages to the waiter traceback .
 time_expired._what += "\n" + filter_log_by(res, filter_by="error=")
 
 In case not found , nothing added and raised exception raises as is.